### DISCUSSION TITLE
keycloak login page customization

### DISCUSSION THREAD

**Discussion by harsh2077** [keycloak/keycloak#34179] [Q&A] [answered]
I have downloaded Keycloak directly from GitHub, and I would like to customize the Keycloak login page. However, I am finding it challenging to identify which files need to be edited due to the numerous files in the Keycloak project. Could you provide guidance on which specific files should be modified for this customization?

**Comment by danielFesenmeyer** [comment]
It depends on what you have to customize ;)

If you are not sure what to customize, the most pragmatic approach is to temporarily add some additional content (maybe with some crazy styling like :fire: to make sure you recognize it) to each template that might need customization.

But consider that your customization might not work in future Keycloak versions. So you should make sure that you have proper tests for your customization.

The best means to customize the Login Pages is via CSS, and, if you only have small changes in logic, via Javascript DOM manipulation.

Depending on your customization requirements, this project might also be interesting: https://www.keycloakify.dev/

**Reply by danielFesenmeyer** [reply, marked as answer]
Btw, the standard login template file is `login.ftl`, probably useful as a starting point.

**Reply by harsh2077** [reply]
@danielFesenmeyer 
Can u help in one more thing 
I have used the following command to build my Keycloak project:
`./mvnw -pl quarkus/deployment,quarkus/dist -am -DskipTests clean install`
Subsequently, I ran the project in development mode with the following command:
`java -jar quarkus/server/target/lib/quarkus-run.jar start-dev`
However, I have noticed that changes made to the source files are not being reflected on the server automatically. Could you please assist me in resolving this issue so that the server reflects changes without needing a manual restart?

**Reply by danielFesenmeyer** [reply]
@harsh2077
Have you just adjusted the built-in Keycloak theme or created your own, inheriting the Keycloak theme?
If you did the first, changes won't be automatically reflected, because the files are delivered from an archive.

For development purposes, it makes sense to use a folder for your theme. (We use this approach even for production.)
You can find more information in the docs: https://www.keycloak.org/docs/26.0.1/server_development/#deploying-themes
